server: TestUpgradeHappensAfterMigrations failed

server.TestUpgradeHappensAfterMigrations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3297360&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3297360&tab=artifacts#/) on master @ [70183840a80160a652af8be84efd66430e4bc96e](https://github.com/cockroachdb/cockroach/commits/70183840a80160a652af8be84efd66430e4bc96e):

Fatal error:

```
panic: cannot do timestamp checking for EngineIterator [recovered]
	panic: cannot do timestamp checking for EngineIterator [recovered]
	panic: cannot do timestamp checking for EngineIterator [recovered]
	panic: cannot do timestamp checking for EngineIterator

```
Stack: 

```
goroutine 188688 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0092361b0, 0x912c258, 0xc00a27ce10)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x15d
panic(0x5b9cae0, 0x905e8a0)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0092361b0, 0x912c258, 0xc00a5926c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x15d
panic(0x5b9cae0, 0x905e8a0)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send.func1(0xc006f5e510, 0xc006f5e5e8, 0xc001d1d000, 0xc006f5e5e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:85 +0x745
panic(0x5b9cae0, 0x905e8a0)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.spanSetReader.NewEngineIterator(0x9196308, 0xc0057ea800, 0xc0058e32c0, 0x0, 0x169acaa16fcb1fce, 0x0, 0x0, 0xc008fe9f58, 0x5, 0x8, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/batch.go:470 +0x1e5
github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval.ScanInterleavedIntents(0x912c258, 0xc00a592720, 0x91964a8, 0xc0070c9d40, 0x9270ac8, 0xc0092601c0, 0x169acaa16fcb1fce, 0x0, 0x0, 0x100000001, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/cmd_scan_interleaved_intents.go:50 +0x1c2
github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateCommand(0x912c258, 0xc00a592720, 0x0, 0x0, 0x0, 0x926bf20, 0xc0070c9d40, 0x9270ac8, 0xc0092601c0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:509 +0x715
github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateBatch(0x912c258, 0xc00a592720, 0x0, 0x0, 0x926bf20, 0xc0070c9d40, 0x9270ac8, 0xc0092601c0, 0x0, 0xc0063fd520, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:271 +0x4f5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatchWithServersideRefreshes(0xc004bc7500, 0x912c258, 0xc00a592720, 0x926bf20, 0xc0070c9d40, 0x9270ac8, 0xc0092601c0, 0xc0063fd520, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_read.go:291 +0x2b4
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatch(0xc004bc7500, 0x912c258, 0xc00a592720, 0xc0063fd520, 0xc00520f180, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_read.go:85 +0x631
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc004bc7500, 0x912c258, 0xc00a592720, 0xc0063fd520, 0x6906d10, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:377 +0x487
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc004bc7500, 0x912c258, 0xc00a5926f0, 0x1, 0xc0063fd520, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:103 +0x349
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:43
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc001d1d000, 0x912c258, 0xc00a5926c0, 0x169acaa16fcb1fce, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:180 +0x5b3
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).Send(0xc008c95f80, 0x912c258, 0xc00a5926c0, 0x169acaa16fcb1fce, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:191 +0x198
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1(0x912c258, 0xc00a5926c0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:905 +0x27e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc0092361b0, 0x912c258, 0xc00a5926c0, 0x62ceaf7, 0x10, 0xc006f5ec80, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:324 +0x12a
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal(0xc0069c5800, 0x912c258, 0xc00a5926c0, 0xc0063fd380, 0x0, 0x0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:893 +0x267
github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch(0xc0069c5800, 0x912c258, 0xc00a592690, 0xc0063fd380, 0xc006f5ef08, 0x5b05dc, 0xc004dfc400)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:986 +0x125
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:469
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch(0xc005eba880, 0x912c258, 0xc00a592690, 0x1, 0x917c528, 0xc0073aa530, 0x169acaa16fcb1fce, 0x0, 0x0, 0x100000001, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:195 +0x25d
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext(0xc005eba880, 0x912c1b0, 0xc004dfc400, 0x169acaa16fcb1fce, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:177 +0x25f
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.raceTransport.SendNext(0x9176128, 0xc005eba880, 0x912c1b0, 0xc004dfc400, 0x169acaa16fcb1fce, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport_race.go:78 +0x443
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas(0xc0012fa000, 0x912c1b0, 0xc004dfc400, 0x169acaa16fcb1fce, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1931 +0xb5b
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch(0xc0012fa000, 0x912c1b0, 0xc004dfc400, 0x169acaa16fcb1fce, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1526 +0x437
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges(0xc0012fa000, 0x912c1b0, 0xc004dfc400, 0x169acaa16fcb1fce, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1144 +0x1f0a
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send(0xc0012fa000, 0x912c1b0, 0xc004dfc400, 0x169acaa16fcb1fce, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:778 +0x7bb
github.com/cockroachdb/cockroach/pkg/kv.(*CrossRangeTxnWrapperSender).Send(0xc008c95e58, 0x912c1b0, 0xc004dfc400, 0x169acaa16fcb1fce, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:220 +0xf5
github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender(0xc008c95e00, 0x912c1b0, 0xc004dfc400, 0x169acaa16fcb1fce, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:875 +0x175
github.com/cockroachdb/cockroach/pkg/kv.(*DB).send(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:858
github.com/cockroachdb/cockroach/pkg/kv.sendAndFill(0x912c1b0, 0xc004dfc400, 0xc006f63280, 0xc005631900, 0x5, 0xc0059b0980)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:786 +0x1bb
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Run(0xc008c95e00, 0x912c1b0, 0xc004dfc400, 0xc005631900, 0xc007cc1020, 0x4a196c)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:809 +0xd7
github.com/cockroachdb/cockroach/pkg/kv.(*DB).ScanInterleavedIntents(0xc008c95e00, 0x912c1b0, 0xc004dfc400, 0x5ff66e0, 0xc007cc1008, 0x5ff66e0, 0xc007cc1020, 0x169acaa16fcb1fce, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:736 +0x154
github.com/cockroachdb/cockroach/pkg/migration/migrations.(*migrateLockTablePool).attemptMigrateRequest(0xc008025c20, 0x912c1b0, 0xc004dfc400, 0xc0067d5bd0, 0x2, 0xc001394e01, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/migration/migrations/separated_intents.go:198 +0x477
github.com/cockroachdb/cockroach/pkg/migration/migrations.(*migrateLockTablePool).run(0xc008025c20, 0x912c1b0, 0xc004dfc400, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/migration/migrations/separated_intents.go:297 +0x296
github.com/cockroachdb/cockroach/pkg/migration/migrations.(*migrateLockTablePool).runMigrateRequestsForRanges.func1(0x912c258, 0xc00a27ce10)
	/go/src/github.com/cockroachdb/cockroach/pkg/migration/migrations/separated_intents.go:110 +0x5f
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc0092361b0, 0x912c258, 0xc00a27ce10, 0xc009ffe000, 0x0, 0xc008cbffe0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0x162
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x28e
```
<details><summary>Log preceding fatal error</summary>
<p>

```
I210813 06:43:29.309477 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1122  executing bump-cluster-version=21.1-116 on nodes n{1}
I210813 06:43:29.311834 188628 server/migration.go:149  [n1,bump-cluster-version] 1123  active cluster version setting is now 21.1-116 (up from 21.1-115(fence))
I210813 06:43:29.313920 188158 migration/migrationmanager/manager.go:131  [n1,intExec=set-version,migration-mgr] 1124  stepping through 21.1-118
I210813 06:43:29.341714 188158 jobs/registry.go:342  [n1,intExec=set-version,migration-mgr] 1125  scheduled jobs [684086425400737793]
I210813 06:43:29.353417 188620 jobs/adopt.go:192  [-] 1126  job 684086425400737793: resuming execution
I210813 06:43:29.363743 188553 jobs/registry.go:1155  [n1] 1127  MIGRATION job 684086425400737793: stepping through state running with error: <nil>
I210813 06:43:29.425918 188553 jobs/registry.go:1155  [n1] 1128  MIGRATION job 684086425400737793: stepping through state succeeded with error: <nil>
I210813 06:43:29.466461 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1129  executing bump-cluster-version=21.1-117(fence) on nodes n{1}
I210813 06:43:29.469177 188587 server/migration.go:149  [n1,bump-cluster-version] 1130  active cluster version setting is now 21.1-117(fence) (up from 21.1-116)
I210813 06:43:29.472339 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1131  executing validate-cluster-version=21.1-118 on nodes n{1}
I210813 06:43:29.476467 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1132  executing bump-cluster-version=21.1-118 on nodes n{1}
I210813 06:43:29.478984 188603 server/migration.go:149  [n1,bump-cluster-version] 1133  active cluster version setting is now 21.1-118 (up from 21.1-117(fence))
I210813 06:43:29.481046 188158 migration/migrationmanager/manager.go:131  [n1,intExec=set-version,migration-mgr] 1134  stepping through 21.1-120
I210813 06:43:29.510004 188158 jobs/registry.go:342  [n1,intExec=set-version,migration-mgr] 1135  scheduled jobs [684086425948258305]
I210813 06:43:29.521085 188644 jobs/adopt.go:192  [-] 1136  job 684086425948258305: resuming execution
I210813 06:43:29.530138 188646 jobs/registry.go:1155  [n1] 1137  MIGRATION job 684086425948258305: stepping through state running with error: <nil>
I210813 06:43:29.601468 188646 jobs/registry.go:1155  [n1] 1138  MIGRATION job 684086425948258305: stepping through state succeeded with error: <nil>
I210813 06:43:29.665718 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1139  executing bump-cluster-version=21.1-119(fence) on nodes n{1}
I210813 06:43:29.668306 188662 server/migration.go:149  [n1,bump-cluster-version] 1140  active cluster version setting is now 21.1-119(fence) (up from 21.1-118)
I210813 06:43:29.671371 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1141  executing validate-cluster-version=21.1-120 on nodes n{1}
I210813 06:43:29.675869 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1142  executing bump-cluster-version=21.1-120 on nodes n{1}
I210813 06:43:29.678006 188742 server/migration.go:149  [n1,bump-cluster-version] 1143  active cluster version setting is now 21.1-120 (up from 21.1-119(fence))
I210813 06:43:29.680129 188158 migration/migrationmanager/manager.go:131  [n1,intExec=set-version,migration-mgr] 1144  stepping through 21.1-122
I210813 06:43:29.707212 188158 jobs/registry.go:342  [n1,intExec=set-version,migration-mgr] 1145  scheduled jobs [684086426600669185]
I210813 06:43:29.718610 188576 jobs/adopt.go:192  [-] 1146  job 684086426600669185: resuming execution
I210813 06:43:29.726556 188710 jobs/registry.go:1155  [n1] 1147  MIGRATION job 684086426600669185: stepping through state running with error: <nil>
I210813 06:43:29.786094 188710 jobs/registry.go:1155  [n1] 1148  MIGRATION job 684086426600669185: stepping through state succeeded with error: <nil>
I210813 06:43:29.824537 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1149  executing bump-cluster-version=21.1-121(fence) on nodes n{1}
I210813 06:43:29.827139 188654 server/migration.go:149  [n1,bump-cluster-version] 1150  active cluster version setting is now 21.1-121(fence) (up from 21.1-120)
I210813 06:43:29.830876 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1151  executing validate-cluster-version=21.1-122 on nodes n{1}
I210813 06:43:29.836059 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1152  executing bump-cluster-version=21.1-122 on nodes n{1}
I210813 06:43:29.838474 188657 server/migration.go:149  [n1,bump-cluster-version] 1153  active cluster version setting is now 21.1-122 (up from 21.1-121(fence))
I210813 06:43:29.840620 188158 migration/migrationmanager/manager.go:131  [n1,intExec=set-version,migration-mgr] 1154  stepping through 21.1-124
I210813 06:43:29.842343 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1155  executing bump-cluster-version=21.1-123(fence) on nodes n{1}
I210813 06:43:29.844802 188728 server/migration.go:149  [n1,bump-cluster-version] 1156  active cluster version setting is now 21.1-123(fence) (up from 21.1-122)
I210813 06:43:29.848054 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1157  executing validate-cluster-version=21.1-124 on nodes n{1}
I210813 06:43:29.852192 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1158  executing bump-cluster-version=21.1-124 on nodes n{1}
I210813 06:43:29.854444 188751 server/migration.go:149  [n1,bump-cluster-version] 1159  active cluster version setting is now 21.1-124 (up from 21.1-123(fence))
I210813 06:43:29.856472 188158 migration/migrationmanager/manager.go:131  [n1,intExec=set-version,migration-mgr] 1160  stepping through 21.1-126
I210813 06:43:29.858264 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1161  executing bump-cluster-version=21.1-125(fence) on nodes n{1}
I210813 06:43:29.860446 188753 server/migration.go:149  [n1,bump-cluster-version] 1162  active cluster version setting is now 21.1-125(fence) (up from 21.1-124)
I210813 06:43:29.863131 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1163  executing validate-cluster-version=21.1-126 on nodes n{1}
I210813 06:43:29.867593 188158 migration/migrationcluster/cluster.go:109  [n1,intExec=set-version,migration-mgr] 1164  executing bump-cluster-version=21.1-126 on nodes n{1}
I210813 06:43:29.869819 188771 server/migration.go:149  [n1,bump-cluster-version] 1165  active cluster version setting is now 21.1-126 (up from 21.1-125(fence))
I210813 06:43:29.872074 188158 migration/migrationmanager/manager.go:131  [n1,intExec=set-version,migration-mgr] 1166  stepping through 21.1-128
I210813 06:43:29.900984 188158 jobs/registry.go:342  [n1,intExec=set-version,migration-mgr] 1167  scheduled jobs [684086427229487105]
I210813 06:43:29.911888 188733 jobs/adopt.go:192  [-] 1168  job 684086427229487105: resuming execution
I210813 06:43:29.921309 188760 jobs/registry.go:1155  [n1] 1169  MIGRATION job 684086427229487105: stepping through state running with error: <nil>
I210813 06:43:29.922524 187136 1@gossip/gossip.go:1505  [n1] 1170  node has connected to cluster via gossip
I210813 06:43:29.923432 187136 kv/kvserver/stores.go:269  [n1] 1171  wrote 0 node addresses to persistent storage
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestUpgradeHappensAfterMigrations PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/server andy-kimball
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUpgradeHappensAfterMigrations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/batch_spanset_test.go
+++ b/pkg/kv/kvserver/batch_spanset_test.go
@@ -385,7 +385,9 @@ func TestSpanSetIteratorTimestamps(t *testing.T) {
 		Key: roachpb.Key("c"), EndKey: roachpb.Key("e")}, hlc.Timestamp{WallTime: 2})
 
 	k1, v1 := storage.MakeMVCCMetadataKey(roachpb.Key("b")), []byte("b-value")
+	k1e := storage.EngineKey{Key: k1.Key}
 	k2, v2 := storage.MakeMVCCMetadataKey(roachpb.Key("d")), []byte("d-value")
+	k2e := storage.EngineKey{Key: k2.Key}
 
 	// Write values that we can try to read later.
 	if err := eng.PutUnversioned(k1.Key, v1); err != nil {
@@ -429,7 +431,7 @@ func TestSpanSetIteratorTimestamps(t *testing.T) {
 		}
 	}()
 
-	{
+	func() {
 		// When accessing at t=2, we're only able to read through the latch declared at t=2.
 		iter := batchAt2.NewMVCCIterator(storage.MVCCKeyAndIntentsIterKind, storage.IterOptions{UpperBound: roachpb.KeyMax})
 		defer iter.Close()
@@ -446,7 +448,7 @@ func TestSpanSetIteratorTimestamps(t *testing.T) {
 		if !reflect.DeepEqual(iter.Key(), k2) {
 			t.Fatalf("expected key %s, got %s", k2, iter.Key())
 		}
-	}
+	}()
 
 	for _, batch := range []storage.Batch{batchAt3, batchNonMVCC} {
 		// When accessing at t=3, we're unable to read through any of the declared latches.
@@ -464,6 +466,57 @@ func TestSpanSetIteratorTimestamps(t *testing.T) {
 			t.Fatalf("expected invalid iterator; found valid at key %s", iter.Key())
 		}
 	}
+
+	// The behavior is the same as above for an EngineIterator.
+	func() {
+		// When accessing at t=1, we're able to read through latches declared at t=1 and t=2.
+		iter := batchAt1.NewEngineIterator(storage.IterOptions{UpperBound: roachpb.KeyMax})
+		defer iter.Close()
+
+		if ok, err := iter.SeekEngineKeyGE(k1e); !ok {
+			t.Fatalf("expected valid iterator, err=%v", err)
+		}
+		engineKey, err := iter.EngineKey()
+		if err != nil {
+			t.Fatalf("expected no error, got err=%v", err)
+		}
+		if !reflect.DeepEqual(engineKey, k1e) {
+			t.Fatalf("expected key %s, got %s", k1e, engineKey)
+		}
+
+		if ok, err := iter.NextEngineKey(); !ok {
+			t.Fatalf("expected valid iterator, err=%v", err)
+		}
+		engineKey, err = iter.EngineKey()
+		if err != nil {
+			t.Fatalf("expected no error, got err=%v", err)
+		}
+		if !reflect.DeepEqual(engineKey, k2e) {
+			t.Fatalf("expected key %s, got %s", k2e, engineKey)
+		}
+	}()
+
+	func() {
+		// When accessing at t=2, we're only able to read through the latch declared at t=2.
+		iter := batchAt2.NewEngineIterator(storage.IterOptions{UpperBound: roachpb.KeyMax})
+		defer iter.Close()
+
+		if ok, _ := iter.SeekEngineKeyGE(k1e); ok {
+			engineKey, err := iter.EngineKey()
+			t.Fatalf("expected invalid iterator; found valid at key %s, err=%v", engineKey, err)
+		}
+
+		if ok, err := iter.SeekEngineKeyGE(k2e); !ok {
+			t.Fatalf("expected valid iterator, err=%v", err)
+		}
+		engineKey, err := iter.EngineKey()
+		if err != nil {
+			t.Fatalf("expected no error, got err=%v", err)
+		}
+		if !reflect.DeepEqual(engineKey, k2e) {
+			t.Fatalf("expected key %s, got %s", k2e, engineKey)
+		}
+	}()
 }
 
 func TestSpanSetNonMVCCBatch(t *testing.T) {

--- a/pkg/kv/kvserver/spanset/batch.go
+++ b/pkg/kv/kvserver/spanset/batch.go
@@ -18,6 +18,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/cockroachdb/pebble"
@@ -234,8 +235,10 @@ func (i *MVCCIterator) SupportsPrev() bool {
 // EngineIterator wraps a storage.EngineIterator and ensures that it can
 // only be used to access spans in a SpanSet.
 type EngineIterator struct {
-	i     storage.EngineIterator
-	spans *SpanSet
+	i         storage.EngineIterator
+	spans     *SpanSet
+	spansOnly bool
+	ts        hlc.Timestamp
 }
 
 // Close is part of the storage.EngineIterator interface.
@@ -249,7 +252,12 @@ func (i *EngineIterator) SeekEngineKeyGE(key storage.EngineKey) (valid bool, err
 	if !valid {
 		return valid, err
 	}
-	if err = i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: key.Key}); err != nil {
+	if key.IsMVCCKey() && !i.spansOnly {
+		mvccKey, _ := key.ToMVCCKey()
+		if err := i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: mvccKey.Key}, i.ts); err != nil {
+			return false, err
+		}
+	} else if err = i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: key.Key}); err != nil {
 		return false, err
 	}
 	return valid, err
@@ -261,7 +269,12 @@ func (i *EngineIterator) SeekEngineKeyLT(key storage.EngineKey) (valid bool, err
 	if !valid {
 		return valid, err
 	}
-	if err = i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{EndKey: key.Key}); err != nil {
+	if key.IsMVCCKey() && !i.spansOnly {
+		mvccKey, _ := key.ToMVCCKey()
+		if err := i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: mvccKey.Key}, i.ts); err != nil {
+			return false, err
+		}
+	} else if err = i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{EndKey: key.Key}); err != nil {
 		return false, err
 	}
 	return valid, err
@@ -332,7 +345,13 @@ func (i *EngineIterator) checkKeyAllowed() (valid bool, err error) {
 	if err != nil {
 		return false, err
 	}
-	if err = i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: key.Key}); err != nil {
+	if key.IsMVCCKey() && !i.spansOnly {
+		mvccKey, _ := key.ToMVCCKey()
+		if err := i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: mvccKey.Key}, i.ts); err != nil {
+			// Invalid, but no error.
+			return false, nil // nolint:returnerrcheck
+		}
+	} else if err = i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: key.Key}); err != nil {
 		// Invalid, but no error.
 		return false, nil // nolint:returnerrcheck
 	}
@@ -467,11 +486,14 @@ func (s spanSetReader) NewMVCCIterator(
 
 func (s spanSetReader) NewEngineIterator(opts storage.IterOptions) storage.EngineIterator {
 	if !s.spansOnly {
-		panic("cannot do timestamp checking for EngineIterator")
+		log.Warningf(context.Background(),
+			"cannot do strict timestamp checking of EngineIterator, resorting to best effort")
 	}
 	return &EngineIterator{
-		i:     s.r.NewEngineIterator(opts),
-		spans: s.spans,
+		i:         s.r.NewEngineIterator(opts),
+		spans:     s.spans,
+		spansOnly: s.spansOnly,
+		ts:        s.ts,
 	}
 }
 

